src/main/java/org/myrobotlab/service/Ros.java

```diff
@@ -1,38 +1,316 @@
+   *   service_response - a service response
+   *      </pre>
+   * 
+   * @author GroG
+   *
+   */
+
+  static public class RosMsg {
+    public List<Object> args;
+    public String compression;
+    public String id; /* optional id */
+    public Object msg;
+    public String op; // publish | subscribe | ? call_service ?
+    public String service;
+    public String topic;
+    /**
+     * return (optional) from service_call
+     */
+    public Object values;
+
+//    public String toString() {
+//      return CodecUtils.toJson(this);
+//    }
+  }
 
-  public final static Logger log = LoggerFactory.getLogger(Ros.class.getCanonicalName());
+  public class RosServiceCallback {
+    public String id;
+    public Object msg;
+  }
+
+  public final static Logger log = LoggerFactory.getLogger(Ros.class);
+  private static final long serialVersionUID = 1L;
 
   public static void main(String[] args) {
-    LoggingFactory.init(Level.WARN);
+    LoggingFactory.init(Level.INFO);
     try {
       Ros ros = (Ros) Runtime.start("ros", "Ros");
-      ros.startService();
-      /*
-       * SwingGui gui = new SwingGui("gui"); gui.startService();
-       */
+
+      WebGui webgui = (WebGui) Runtime.create("webgui", "WebGui");
+      // webgui.setSsl(true);
+      webgui.autoStartBrowser(false);
+      webgui.setPort(8888);
+      // webgui.setSsl(true);
+      webgui.startService();
 
     } catch (Exception e) {
-      Logging.logError(e);
+      log.error("main threw", e);
     }
   }
 
+  Map<String, RosServiceCallback> callbacks = new HashMap<>();
+
+  @SuppressWarnings("rawtypes")
+  transient private WsClient client = null;
+
+  protected boolean connected = false;
+
   public Ros(String n, String id) {
     super(n, id);
   }
 
-  static public String[] getCategories() {
-    return new String[] { "bridge" };
+  @Override
+  public ServiceConfig apply(ServiceConfig c) {
+    RosConfig config = (RosConfig) super.apply(c);
+    if (config.connect) {
+      connect(config.bridgeUrl);
+      if (config.subscriptions != null) {
+        for (String topic : config.subscriptions) {
+          rosSubscribe(topic);
+        }
+      }
+    }
+    return c;
+  }
+
+  // FIXME - TODO reconnect
+  public void connect(String url) {
+
+    if (connected) {
+      info("already connected");
+      return;
+    }
+
+    client = new WsClient();
+    client.connect(this, url);
+  }
+
+  // TODO - getTopics
+  // getSubscriptions
+  // setNodeName ?
+  //
+
+  public void disconnect() {
+    if (client != null) {
+      client.close();
+    }
+  }
+
+  @Override
+  public ServiceConfig getConfig() {
+    RosConfig config = (RosConfig) super.getConfig();
+    config.connect = connected;
+    return config;
+  }
+
+  public List<String> getSubscriptions() {
+    return ((RosConfig) config).subscriptions;
+  }
+
+  public List<String> getTopics() throws InterruptedException {
+
+    RosServiceCallback ret = rosCallService("/rosapi/topics");

```

Comment:
With some generics trickery, we can leverage the type system for ros callbacks and messages. I managed to get this section to look something like this:
```java
public List<String> getTopics() {

    RosServiceCallback<Map<String, List<String>>> ret =
            rosCallService("/rosapi/topics", new TypeReference<>() {});
    if (ret == null) {
      return null;
    }
    if (ret.msg != null) {
      List<String> topics = ret.msg.values.get("topics");
      log.info(ret.toString());
      return topics;
    } else {
      return new ArrayList<>();
    }
  }
  }
```

Notice the `new TypeReference<>(){}` part, that's a trick in Java that lets us manually reify generic type information, and that information will be available at runtime. `TypeReference` is a Jackson class but Gson has an equivalent called `TypeToken`. The advantage of this mechanism is (with an enhancement to CodecUtils) we can pass this type information to the deserializers and be able to deserialize RosMsg to the expected return type (in my code RosServiceCallback has a generic RosMsg msg instead of an Object). Such a mechanism allows us to leverage Java's static type system for type checking without the need for casting while simultaneously ensuring an incorrectly formatted RosMsg throws an error at deserialization rather than a cast exception later on in the code. This should be perfectly fine to do with Ros because of its strict message format

Reply:
heh part of what i don't like about ros is its brittle types and the painful catkin_ws  .. but it could be advantageous to get back type info ... I don't think i want to do it in this pr though